lib/src/LLM/context_generator.dart

```diff
@@ -9,14 +16,26 @@ Map<Declaration?, List<Declaration>> buildDependencyContext(
   Declaration declaration, {
   int maxDepth = 1,
 }) {
-  final visitedDeclarations = <Declaration>{};
-  final parentMap = <Declaration?, List<Declaration>>{};
+  final parentMap = <Declaration?, Set<Declaration>>{};
 
   for (final dependency in declaration.dependsOn) {
-    _dfs(dependency, visitedDeclarations, parentMap, maxDepth: maxDepth);
+    _dfs(dependency, parentMap, maxDepth: maxDepth);
+  }
+
+  // Remove any declarations from the top-level list that are also keys.
+  final topLevel = parentMap[null]?.toList() ?? <Declaration>[];
```

Comment:
I'm guessing you're not directly iterating `parentMap[null]` because you want to avoid modification during iteration? In that case a simpler/faster approach is to use [`removeWhere`](https://api.dart.dev/dart-core/Set/removeWhere.html): `parentMap[null]?.removeWhere(...)`.

In fact, if you switched `parentMap`'s values to `Set<Declaration>` to make `remove` fast in this loop, `removeWhere` is also defined for lists, so you could switch back.